### PR TITLE
Restore original params for pruners in AllenNLP integration

### PR DESCRIPTION
**Pull request by himkt**
Ref. https://github.com/himkt/allennlp-optuna/issues/39

### 1. `SuccessiveHalvingPruner`

`SuccessiveHalvingPruner` doesn't save `min_resource` parameter when it is `"auto"`.
ref. https://github.com/optuna/optuna/blob/853b94b459ecf9e434fb1e28dda7f07349eec20c/optuna/pruners/_successive_halving.py#L157-L158

However, `AllenNLPExecutor` doesn't take care of it.
ref. https://github.com/optuna/optuna/blob/853b94b459ecf9e434fb1e28dda7f07349eec20c/optuna/integration/allennlp.py#L169

This PR update `min_resource` to `"auto"` when `pruner._min_resource` is `None`.

### 2. `ThresholdPruner`

`SuccessiveHalvingPruner` internally updates `lower` or `upper` before saving them in the initializer.
I also add restoring logics for them.

ref. https://github.com/optuna/optuna/blob/853b94b459ecf9e434fb1e28dda7f07349eec20c/optuna/pruners/_threshold.py#L95-L96

## Discussion

I'm not sure whether it would be better to change the initialization logic of `SuccessiveHalvingPruner` and `ThresholdPruner`. I'd other developers' opinion.

## Description of the changes
- 2d002cc updates `min_resource` when `pruner._min_resource` is `None`
- aaaf60f updates `lower` or `upper` when `pruner._lower` or `pruner._upper` is `None`, respectively
- 6b5ade5 adds test cases
- ef9370d fixes style

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
I think this PR is related to #2539. If all pruners provide the `__repr__` methods, the AllenNLP integration can use them to reproduce the pruner instances. In addition, the pruners' arguments are literals and we can introduce `__repr__` in a straightforward way.

What do you think of it?

**Comment by hvy** [comment]
@toshihikoyanase , @Crissman are you able to take a look at this PR (after the implementation of `__repr__`) ?

**Comment by himkt** [comment]
@toshihikoyanase Thank you for your interest! I'd like to know what you think in detail.
Do you suggest that implements `__repr__` and uses `eval` to restore it?

**Comment by toshihikoyanase** [comment]
> Do you suggest that implements `__repr__` to show a class name and its arguments and uses `eval` to restore it? (just to confirm)

Yes, I do. If `__repr__` is available, the `_fetch_pruner_config` method can be simpler. For example, instead of adding each `__init__` parameters to `kwargs`, we can add f`{pruner}` to it. But, it can break the backward compatibility, and we'll need more detailed design discussion.

**Review by Crissman — COMMENTED** [review]
Thanks for the PR. Some minor comments.

File: optuna/integration/allennlp.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ def wrapper(subclass: Any, *args: Any, **kwargs: Any) -> Any:
             return wrapper
 
 
+_NONE = "<PYTHON_NONE_OBJECT>"
```
[the remark this answers is not in the document]
Comment (himkt) [reply optuna/integration/allennlp.py]: This dirty hack is removed in d3a7d97891e6fa0d0450dd484a8b2ce351daf568.

**Comment by himkt** [comment]
@toshihikoyanase 
I tried to implement `__repr__` and restoring logic for a pruner based on `__repr__`.
I want your feedback.

https://github.com/himkt/optuna/compare/ef9370d1fadbbb7bb72abb206852b45f66c6fab9...d3a7d97891e6fa0d0450dd484a8b2ce351daf568

**Comment by Crissman** [comment]
@himkt Please have a look at the mypy error from the changes to pruners/_base.py 🙇

**Comment by himkt** [comment]
@Crissman Sorry for the delay, I created another PR in #2707. Let me close this PR and create a new PR after merging 2707.

**Pull request closed without merging**